File: app/src/main/java/com/szymkowski/personaltrainercompanion/OverviewActivity.java
package com.szymkowski.personaltrainercompanion;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import com.getbase.floatingactionbutton.FloatingActionButton;
import com.getbase.floatingactionbutton.FloatingActionsMenu;
import com.szymkowski.personaltrainercompanion.core.RepositoryCallback;
import com.szymkowski.personaltrainercompanion.payments.AddPaymentDialog;
import com.szymkowski.personaltrainercompanion.payments.AddPaymentDialogCallback;
import com.szymkowski.personaltrainercompanion.payments.domain.PaymentDTO;
import com.szymkowski.personaltrainercompanion.payments.domain.PaymentRepository;
import com.szymkowski.personaltrainercompanion.trainings.domain.TrainingDTO;
import com.szymkowski.personaltrainercompanion.trainings.domain.TrainingsRepository;

import org.joda.time.DateTime;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

import java.util.Locale;

public class OverviewActivity extends AppCompatActivity implements AddPaymentDialogCallback, RepositoryCallback {

    private static final String TAG = OverviewActivity.class.getSimpleName();
    private TextView mLastPaymentInfoText;

    private PaymentRepository mPaymentRepository;
    private TrainingsRepository mTrainingsRepository;
    private TextView mNumberOfTrainingsInfoText;
    private TextView mLastTrainingInfoText;
    private DateTimeFormatter dateTimeFormatter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_overview);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        dateTimeFormatter = DateTimeFormat.forPattern(getResources().getString(R.string.date_time_format)).withLocale(Locale.getDefault());

        mPaymentRepository = new PaymentRepository(this, this);
        mLastPaymentInfoText = (TextView) findViewById(R.id.last_payment_info);
        mLastTrainingInfoText = (TextView) findViewById(R.id.last_training_info_date);

        mTrainingsRepository = new TrainingsRepository(this,this, mPaymentRepository);
        mNumberOfTrainingsInfoText = (TextView) findViewById(R.id.number_of_trainings_remaining);

        final FloatingActionsMenu floatingActionsMenu = (FloatingActionsMenu) findViewById(R.id.fab_menu);
        FloatingActionButton fabAddPayment = (FloatingActionButton) findViewById(R.id.fab_action_add_payment);
        fabAddPayment.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Dialog addPaymentDialog = new AddPaymentDialog(OverviewActivity.this, OverviewActivity.this);
                addPaymentDialog.show();
                floatingActionsMenu.collapse();
            }
        });
        FloatingActionButton fabAddSingleTraining = (FloatingActionButton) findViewById(R.id.fab_action_add_training);
        fabAddSingleTraining.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                AlertDialog.Builder builder = new AlertDialog.Builder(OverviewActivity.this);
                builder.setTitle(R.string.add_single_training_title);
                builder.setMessage(R.string.add_single_training_message);
                builder.setNegativeButton(android.R.string.no, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                    }
                });
                builder.setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        mTrainingsRepository.addTraining(new TrainingDTO(new DateTime()));
                        onDatasetChanged();
                        dialog.dismiss();
                    }
                });
                builder.show();
            }
        });

        Log.i(TAG, "Activity created");
    }

    @Override
    protected void onResume() {
        super.onResume();
        updateLastPayment();
        updateNumberOfTrainingsRemaining();
        updateLastTrainingInfo();

    }

    private void updateLastTrainingInfo() {
        DateTime dateTime = mTrainingsRepository.getLatestTrainingDate();
        if (dateTime != null) {
            mLastTrainingInfoText.setText(dateTimeFormatter.print(dateTime));
        }
    }

    private void updateNumberOfTrainingsRemaining() {
        mNumberOfTrainingsInfoText.setText(String.format(getResources().getString(R.string.number_format_string), mTrainingsRepository.getNumberOfTrainingsRemaining()));
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_overview, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void addPayment(PaymentDTO newPayment) {
        mPaymentRepository.addPayment(newPayment);
    }

    private void updateLastPayment() {
        PaymentDTO lastPaymentDto = mPaymentRepository.getLastPaymentDTO();
        String paymentInfoText;
        if (lastPaymentDto == null) {
            paymentInfoText = getResources().getString(R.string.no_data_found);
        } else {
            String rawPaymentInfo = getResources().getString(R.string.last_payment_info_string);
            String dateTime = dateTimeFormatter.print(lastPaymentDto.getPaymentDate());
            paymentInfoText = String.format(rawPaymentInfo, dateTime, lastPaymentDto.getNumberOfClassesPaid());
        }
        mLastPaymentInfoText.setText(paymentInfoText);
        updateNumberOfTrainingsRemaining();
    }

    @Override
    public void onDatasetChanged() {
        updateLastPayment();
        updateNumberOfTrainingsRemaining();
        updateLastTrainingInfo();
    }
}

File: app/src/main/java/com/szymkowski/personaltrainercompanion/OverviewActivity.kt
package com.szymkowski.personaltrainercompanion

import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import com.szymkowski.personaltrainercompanion.core.DateFormatter
import com.szymkowski.personaltrainercompanion.core.RepositoryCallback
import com.szymkowski.personaltrainercompanion.payments.AddPaymentDialog
import com.szymkowski.personaltrainercompanion.payments.AddPaymentDialogCallback
import com.szymkowski.personaltrainercompanion.payments.domain.PaymentDTO
import com.szymkowski.personaltrainercompanion.payments.domain.PaymentRepository
import com.szymkowski.personaltrainercompanion.trainings.TrainingsActivity
import com.szymkowski.personaltrainercompanion.trainings.domain.TrainingDTO
import com.szymkowski.personaltrainercompanion.trainings.domain.TrainingsRepository
import kotlinx.android.synthetic.main.activity_overview.*
import kotlinx.android.synthetic.main.content_overview.*
import org.joda.time.DateTime
import org.joda.time.format.DateTimeFormatter

class OverviewActivity : AppCompatActivity(), AddPaymentDialogCallback, RepositoryCallback {


    lateinit private var mPaymentRepository: PaymentRepository
    lateinit private var mTrainingsRepository: TrainingsRepository
    lateinit private var dateTimeFormatter: DateTimeFormatter


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_overview)
        setSupportActionBar(toolbar)

        dateTimeFormatter = DateFormatter.getDateFormatter(resources.getString(R.string.date_time_format))

        mPaymentRepository = PaymentRepository(this, this)

        mTrainingsRepository = TrainingsRepository(this, this, mPaymentRepository)


        fab_action_add_payment.setOnClickListener({
            val addPaymentDialog = AddPaymentDialog(this@OverviewActivity, this@OverviewActivity)
            addPaymentDialog.show()
            fab_menu.collapse()
        })

        fab_action_add_training.setOnClickListener {
            val builder = AlertDialog.Builder(this@OverviewActivity)
            builder.setTitle(R.string.add_single_training_title)
            builder.setMessage(R.string.add_single_training_message)
            builder.setNegativeButton(android.R.string.no) { dialog, which -> dialog.dismiss() }
            builder.setPositiveButton(android.R.string.yes) { dialog, which ->
                mTrainingsRepository.addTraining(TrainingDTO(DateTime()))
                fab_menu.collapse()
                dialog.dismiss()
            }
            builder.show()
        }
        last_training_info_date.setOnClickListener {
            val intent = Intent(this@OverviewActivity, TrainingsActivity::class.java)
            startActivity(intent)

        }

        Log.i(TAG, "Activity created")
    }

    override fun onResume() {
        super.onResume()
        updateLastPayment()
        updateNumberOfTrainingsRemaining()
        updateLastTrainingInfo()

    }

    private fun updateLastTrainingInfo() {
        val dateTime = mTrainingsRepository.latestTrainingDate
        if (dateTime != null) {
            last_training_info_date.text = dateTimeFormatter.print(dateTime)
        }
    }

    private fun updateNumberOfTrainingsRemaining() {
        val trainingsRemaining = mTrainingsRepository.numberOfTrainingsRemaining
        number_of_trainings_remaining.text = trainingsRemaining.toString()

        if (trainingsRemaining > 0) {
            fab_action_add_training.visibility = View.VISIBLE
        } else {
            fab_action_add_training.visibility = View.GONE
        }


    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_overview, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true
        }

        return super.onOptionsItemSelected(item)
    }

    override fun addPayment(newPayment: PaymentDTO) {
        mPaymentRepository.addPayment(newPayment)
    }

    private fun updateLastPayment() {
        val lastPaymentDto = mPaymentRepository.lastPaymentDTO
        val paymentInfoText: String
        if (lastPaymentDto == null) {
            paymentInfoText = resources.getString(R.string.no_data_found)
        } else {
            val rawPaymentInfo = resources.getString(R.string.last_payment_info_string)
            val dateTime = dateTimeFormatter.print(lastPaymentDto.paymentDate)
            paymentInfoText = String.format(rawPaymentInfo, dateTime, lastPaymentDto.numberOfClassesPaid)
        }
        last_payment_info.text = paymentInfoText
        updateNumberOfTrainingsRemaining()
    }

    override fun onDatasetChanged() {
        updateLastPayment()
        updateNumberOfTrainingsRemaining()
        updateLastTrainingInfo()

    }

    companion object {

        private val TAG = OverviewActivity::class.java.simpleName
    }
}
